### PR TITLE
Allow to specify sample page name via --example-page-name

### PR DESCRIPTION
**Pull request by ledestin**
Not sure what to do about the Rubocop warning.

---
This change is [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/869)

### CODE REVIEW

**Comment by ledestin** [comment]
@justin808 hey. I've added tests as well, but haven't updated docs.

The build fails because CI lacks `redux`, it all happened in my local dev env, and I had to `yarn add` React, Redux. Why would it happen in CI though?

**Comment by justin808** [comment]
@ledestin Please see our other generator tests. If you specify the --redux option, the generator should be adding redux to the package.json.

**Comment by ledestin** [comment]
@justin808 well, CI passes, please review. I'll update docs later.

**Comment by justin808** [comment]
@ledestin Please keep going with the docs, and please review your changes in case you want to add any commentary. I have to give a talk tomorrow and after that I can review this and get it merged ASAP.

**Comment by ledestin** [comment]
@justin808 updated docs. What kind of commentary do you have in mind?

One thing I'd like to add is that it's not optimal for development of the example page to have source code files as templates. I feel it'd be easy to have `HelloWorldApp` instead of `<%= example_page_name %>App`, but templates are a common way of doing it. I could potentially replace `HelloWorld` without templates though. What do you think?

**Comment by justin808** [comment]
@ledestin Is this one ready to go? I'm about ready to merge and release as 8.1.0. Please update your branch on top master. Best to rebase to one commit on top of master.

**Comment by ledestin** [comment]
@justin808 it's ready

**Comment by justin808** [comment]
@ledestin I finally got around to this one. 

**REALLY OUTSTANDING!**

I left you some suggestions and one concern regarding one place you may have left in a HelloWorld. Please address my comments and concerns.

And then rebase on top of master.

Thank you again for this **OUTSTANDING PR**!

---

Reviewed 40 of 40 files at r1.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[Gemfile, line 46 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHENNl4gEivL5IyaZC:-KqHENNl4gEivL5IyaZD:bap5o0k) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/Gemfile#L46)):*
> ```Ruby
> 
> group :development do
>   gem "listen"
> ```

@ledestin Why is "listen" needed for this project?

---

*[docs/tutorial.md, line 71 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHEUqXaTDqKkP59BuQ:-KqHEUqXaTDqKkP59BuR:b-oixwo1) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/docs/tutorial.md#L71)):*
> ```Markdown
> 
> You can also specify name for the generated page with `react_on_rails:install
> --example-page-name MyApp`. It will affect URL path (`http://localhost:3000/my_app/`) and source code names (instead of HelloWorld things will be called MyApp).
> ```

How about we shorten this to `install-name`?

`example-page-name` seems too long, and the route is actually the snake case version of this.

@robwise @samnang Name? 

---

*[docs/basics/generator.md, line 15 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHEp8DIZg2l6YDa0_R:-KqHEp8DIZg2l6YDa0_S:b70kbgf) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/docs/basics/generator.md#L15)):*
> ```Markdown
> 
> Options:
>   -E, [--example-page-name=EXAMPLE_PAGE_NAME]      # Name the example page EXAMPLE_PAGE_NAME
> ```

How about we shorten this to `--name` and `-N`? 

or

`--example` and `-En`? 



(we're removing the "node server option in another PR)

`example-page-name` seems too long, and the route is actually the snake case version of this.

@robwise @samnang Name? 

---

*[docs/basics/generator.md, line 17 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHG7xK4B9ihhTshD_a:-KqHG7xK4B9ihhTshD_b:bofcu7d) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/docs/basics/generator.md#L17)):*
> # Install Redux gems and Redux version of Example Page. Default: false

Use

> # Use Redux in the example. Default: false

---

*[lib/generators/react_on_rails/generator_helper.rb, line 34 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHHHby3E0UwcNMaraA:-KqHHHby3E0UwcNMaraB:b-68dx3e) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/lib/generators/react_on_rails/generator_helper.rb#L34)):*
> ```Ruby
>   def create_client_directories(*dir_names)
>     dir_names.each do |dir_name|
>       dest_dir = "client/app/bundles/HelloWorld/#{dir_name}"
> ```

@ledestin Do you still want HelloWorld here?

---

*[lib/generators/react_on_rails/install_generator.rb, line 22 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHHax-vJ7pCVG2zdfm:-KqHHax-vJ7pCVG2zdfn:bywmiwt) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/lib/generators/react_on_rails/install_generator.rb#L22)):*
> desc: "Install Redux gems and Redux version of Example Page. Default: false",

How about

> desc: "Use Redux Version. Default: false",

---

*[lib/generators/react_on_rails/option_helper.rb, line 11 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHHt5u_xeEuSVg_sNe:-KqHHt5u_xeEuSVg_sNf:b-jq260s) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/lib/generators/react_on_rails/option_helper.rb#L11)):*
> ```Ruby
> 
>   def example_page_lower_camelcase
>     options.example_page_name.camelize :lower
> ```

I prefer using parens around arguments.

---

*[lib/generators/react_on_rails/option_helper.rb, line 38 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHITQGz0BMFPOKyVqf:-KqHITQGz0BMFPOKyVqg:b-jgngaj) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/lib/generators/react_on_rails/option_helper.rb#L38)):*
> ```Ruby
>                    type: :string,
>                    default: "HelloWorld",
>                    desc: "Name the example page EXAMPLE_PAGE_NAME",
> ```

see prior comment on using either "example" or "name"

---

*[lib/generators/react_on_rails/templates/base/base/app/views/hello_world/index.html.erb.tt, line 2 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHJ0qtrkiN3WpsCbTx:-KqHJ0qtrkiN3WpsCbTy:bp9ntg4) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/lib/generators/react_on_rails/templates/base/base/app/views/hello_world/index.html.erb.tt#L2)):*
> ```
> <h1><%= example_page_title %></h1>
> <%%= react_component("<%= config[:component_name] %>", props: @<%= example_page_path %>_props, prerender: false) %>
> ```

The "path" suffix is a bit misleading here. You mean the snake_case name.

However, I'm OK with this as-is.

---

*[lib/generators/react_on_rails/templates/base/base/client/webpack.config.js.tt, line 24 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHJCpO2Z5ev7rMX6Nu:-KqHJCpO2Z5ev7rMX6Nv:bqjmhus) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/lib/generators/react_on_rails/templates/base/base/client/webpack.config.js.tt#L24)):*
> ```
>       'es5-shim/es5-sham',
>       'babel-polyfill',
>       './app/bundles/<%= example_page_name %>/startup/registration',
> ```

In this case, we're using the PascalCase name.

Should we validate the param passed to be in PascalCase? I think so.

---

*[lib/generators/react_on_rails/templates/base/base/client/app/bundles/HelloWorld/components/HelloWorld.jsx.tt, line 4 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHJLnB3gDxsJmigyo9:-KqHJLnB3gDxsJmigyoA:bfo5piu) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/lib/generators/react_on_rails/templates/base/base/client/app/bundles/HelloWorld/components/HelloWorld.jsx.tt#L4)):*
> ```
> import React from 'react';
> 
> export default class <%= example_page_name %> extends React.Component {
> ```

OK - starting to think that that example_page_name might be named example_page_name_pascal_case (example_pascal_case or name_pascal_case) and the path name  being example_page_name_snake_case (or example_snake_case or name_snake_case). 

Long term maintainability might be helped by this.

---

*[lib/generators/react_on_rails/templates/redux/base/client/app/bundles/HelloWorld/actions/helloWorldActionCreators.jsx.tt, line 3 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHK4yFrpXgkaJSaKlW:-KqHK4yFrpXgkaJSaKlX:btoci7a) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/lib/generators/react_on_rails/templates/redux/base/client/app/bundles/HelloWorld/actions/helloWorldActionCreators.jsx.tt#L3)):*
> ```
> /* eslint-disable import/prefer-default-export */
> 
> import { <%= example_page_path_in_caps %>_NAME_UPDATE } from '../constants/<%=
> ```

`example_in_caps`

`example_in_snake_case`

`example_in_pascal_case`

???

---

*[spec/react_on_rails/generators/install_generator_spec.rb, line 13 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHKgpnp6chRvGmznty:-KqHKgpnp6chRvGmzntz:b-iraakr) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/spec/react_on_rails/generators/install_generator_spec.rb#L13)):*
> ```Ruby
>   let(:default_example_page_name) { "HelloWorld" }
>   let(:default_example_page_path) { "hello_world" }
>   let(:default_example_page_lower_camelcase) { "helloWorld" }
> ```

We typically don't need to say lower_camelcase for camelcase. However, this is OK.

---

*[spec/react_on_rails/support/generator_spec_helper.rb, line 4 at r1](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-KqHKwmE9XSONikX9oz1:-KqHKwmE9XSONikX9oz2:b-3fn8de) ([raw file](https://github.com/shakacode/react_on_rails/blob/e34abe37d755f7b648331f4046cbe377b320d8fe/spec/react_on_rails/support/generator_spec_helper.rb#L4)):*
> ```Ruby
> 
> require_relative "../simplecov_helper"
> require_relative "../spec_helper"
> ```

why this addition?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/869#-:-KqHLHNan30O3aPgoEkW:bl8qhdr)*

**Comment by justin808** [comment]
See #716.

**Comment by justin808** [comment]
We can reconsider this at some point later. I have not gotten enough demand for customization of the naming to be added to the generator setup.
